components/nodemanager-service/pgdb/project_update.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	db.projectUpdateComplete()
+}
+
+func (db *DB) projectUpdateStatusUpdate(percentageComplete float32) {
+	db.ProjectUpdate.jobStatus = project_update_lib.JobStatus{
+		Completed:             false,
+		PercentageComplete:    percentageComplete,
+		EstimatedEndTimeInSec: 0,
+	}
+}
+
+func (db *DB) projectUpdateComplete() {
+	db.ProjectUpdate.jobStatus = project_update_lib.JobStatus{
+		Completed:             true,
+		PercentageComplete:    1.0,
+		EstimatedEndTimeInSec: 0,
+	}
+	db.ProjectUpdate.running = false
+}
+
+func (db *DB) projectUpdateFail(err error) {
+	db.ProjectUpdate.jobStatusError = err
+	db.ProjectUpdate.jobStatus = project_update_lib.JobStatus{
+		Completed:             true,
+		PercentageComplete:    1.0,
+		EstimatedEndTimeInSec: 0,
+	}
+	db.ProjectUpdate.running = false
+}
+
+// updateNodeProjectIDs - update the nodes project IDs
+func (db *DB) updateNodeProjectIDs(node *NodeProjectData, matchingProjectIDs []string) error {
+	return Transact(db, func(tx *DBTrans) error {
+		err := tx.updateNodeProjects(node.ID, matchingProjectIDs)
+		if err != nil {
+			return errors.Wrap(err, "updateNodeProjectIDs unable to add projects to node")
+		}
+
+		return nil
+	})
+}
+
+func (db *DB) getAllNodes() ([]*NodeProjectData, error) {
+	var nodesDaos []*NodeProjectData
+	_, err := db.Select(&nodesDaos, selectNodesProjectData)
+
+	for _, node := range nodesDaos {
+		dbProjectsData := []*nodes.ProjectsData{}
+		err := json.Unmarshal([]byte(node.Data), &dbProjectsData)
+		if err != nil {
+			return nodesDaos, errors.Wrap(err, "fromDBNode unable to unmarshal projects data")
+		}
+
+		node.ProjectDataKeyValues = dbProjectsData
+	}
+	return nodesDaos, err
+}
+
+func getMatchingProjectIDs(node *NodeProjectData, projectRules map[string]*iam_v2.ProjectRules) []string {
+	matchingProjects := make([]string, 0)
+
+	for projectName, project := range projectRules {
+		if nodeMatchesRules(node, project.Rules) {
+			matchingProjects = append(matchingProjects, projectName)
+		}
+	}
+
+	return matchingProjects
+}
+
+// Only one rule has to be true for the project to match (ORed together).
+func nodeMatchesRules(node *NodeProjectData, rules []*iam_v2.ProjectRule) bool {
+	for _, rule := range rules {
+		if rule.Type == iam_v2.ProjectRuleTypes_NODE && nodeMatchesAllConditions(node, rule.Conditions) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// All the conditions must be true for a rule to be true (ANDed together).
+// If there are no conditions then the rule is false
+func nodeMatchesAllConditions(node *NodeProjectData, conditions []*iam_v2.Condition) bool {
+	if len(conditions) == 0 {
+		return false
+	}
+
+	for _, condition := range conditions {
+		switch condition.Attribute {
+		case iam_v2.ProjectRuleConditionAttributes_CHEF_SERVER:
+			values := node.getValues("chef_server")
+			if len(values) == 0 {
+				return false
+			}
+
+			if !stringutils.SliceContains(condition.Values, values[0]) {
+				return false
+			}
```

Comment:
Consider more concisely:
```
return len(values) > 0 && stringutils.SliceContains(condition.Values, values[0])
```

Reply:
That change does not work because it would return after the first condition. All conditions must be true for the rule to be true. The conditions are 'or'ed together. 

I did combine the two if statements like below. 
`if len(values) == 0 || !stringutils.SliceContains(condition.Values, values[0]) {`